packages/react/src/components/Dropdown/Dropdown.tsx

```diff
@@ -1023,6 +1054,42 @@ class Dropdown extends AutoControlledComponent<Extendable<DropdownProps>, Dropdo
   private isValueEmpty = (value: ShorthandValue | ShorthandCollection) => {
     return _.isArray(value) ? value.length < 1 : !value
   }
+
+  private getHighlightedIndexOnArrowKeyOpen = (
+    changes: StateChangeOptions<ShorthandValue>,
+  ): number => {
+    // if open by ArrowUp, index should change by -1.
+    if (changes.type === Downshift.stateChangeTypes.keyDownArrowUp) {
+      const itemsLength = this.getItemsFilteredBySearchQuery().length
+      if (_.isNumber(this.state.highlightedIndex)) {
+        const newIndex = this.state.highlightedIndex - 1
+        return newIndex < 0 ? itemsLength - 1 : newIndex
+      }
+      return itemsLength - 1
+    }
+    // if open by ArrowDown, index should change by +1.
+    if (changes.type === Downshift.stateChangeTypes.keyDownArrowDown) {
+      const itemsLength = this.getItemsFilteredBySearchQuery().length
+      if (_.isNumber(this.state.highlightedIndex)) {
+        const newIndex = this.state.highlightedIndex + 1
+        return newIndex >= itemsLength ? 0 : newIndex
+      }
+      return 0
```

Comment:
can we encapsulate logic form lines 1072-1077 (and 1063-1068) in smaller methods? logic is complex and because of branching it becomes hard to read

Reply:
there are just 4 lines each, and a separate method with some more if-else to change values depending on that `stateChangeTypes` ... I would rather leave it like this.